### DISCUSSION TITLE
Problems with not linked accounts after upgrading from v3 to v4

### DISCUSSION THREAD

**Discussion by chrishornmem** [nextauthjs/next-auth#3916] [Help]
Hi,

I am experiencing the "OAuthAccountNotLinked" ("To confirm your identity, sign in with the same account you used originally."), error for my users after migrating from v3 to v4.

I am using GoogleProvider.

Seems like all users are affected (on local dev not production :) ) 

I actually don't know how recover from this error, I've tried deleting the sessions in the db but to no avail.

Do you know if this particular migration scenario has been tested?  Any ideas what could be wrong?

I was using mongo database sessions in v3, like this:
```
  // v3 [...nextauth].js
  database: {
    type: 'mongodb',
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useFindAndModify: false,
    url: process.env.DATABASE_URL,
  }
```
and now in v4:
```
// v4[...nextauth].js
{
  adapter: MongoDBAdapter(clientPromise),
  session: {
    strategy: 'database',
    maxAge: 30 * 24 * 60 * 60, 
    updateAge: 24 * 60 * 60,
}
```

**Comment by chrishornmem** [comment]
@balazsorban44 I think I found the problem - in v3 the database model for accounts table uses a field named "providerId" whereas in v4 it seems to have been renamed to"provider".  That means there is a db migration to do when upgrading from v3 to v4?

**Reply by balazsorban44** [reply]
Yes there is as documented here https://next-auth.js.org/getting-started/upgrade-v4#schema-changes 

To easyle the migration process, @ndom91 has put together some help in this PR:

https://github.com/nextauthjs/next-auth/pull/3874

**Reply by ndom91** [reply]
Yeah so that PR with migration instructions is obviously still open.. If you follow it / use the migration script we'd really appreciate any feedback you have about it!

**Reply by chrishornmem** [reply]
@ndom91 I will take a look.  It wasn't obvious when reading the existing Migration guide that the schema changes would actually break the code.  Can we make it clearer in the docs that a database migration must be performed otherwise it's not going to work?

**Reply by chrishornmem** [reply]
@ndom91 I am using mongodb so I can't provide feedback on your SQL migration script.  However, as far as I can tell, there is only one command needed for Mongo migration (since the new fields will be populated automatically and most of the existing fields already had the same names, eg `userId`).

`db.getCollection('accounts').updateMany( {}, { $rename: { "providerId": "provider" } } )`

**Reply by ndom91** [reply]
@chrishornmem Thanks for the quick feedback. You're right, with Mongo being a document database, a lot of the migration becomes unnecessary. However after going through it one more time, I see there are actually a few more renames that would seemingly also have to be done in mongo.

For example:

```sql
db.getCollection('accounts').updateMany({}, {
  $rename: {
    "provider_id": "provider",
    "provider_account_id": "providerAccountId",
    "user_id": "userId",
    "access_token_expires": "expires_at"
  }
})
db.getCollection('users').updateMany({}, {
  $rename: {
    "email_verified": "emailVerified"
  }
})
db.getCollection('sessions').updateMany({}, {
  $rename: {
    "session_token": "sessionToken",
    "user_id": "userId"
  }
})
```

What do you think?

**Reply by chrishornmem** [reply]
@ndom91 Well I just checked on my production system with the following in my package.json

`"next-auth": "^3.29.0",`

And for me, `sessionToken`, `userId` and `emailVerified` are already named correctly.

Here is the full list on **v3:**

**accounts collection**
```compoundId
userId
providerType
providerId  -> beware! must be renamed to `provider` (causes the issue I have mentioned in this ticket) 
providerAccountId
refreshToken
accessToken
accessTokenExpires
```
**sessions collection**
```userId
expires
sessionToken
accessToken
```

**users collection**
```name
email
image
emailVerified
```
Your suggestion won't cause any issues of course.

**Reply by ndom91** [reply]
@chrishornmem great, thanks for double checking! 

It seems the v3 schema that I went off of (I believe it was the latest one in the docs that I could find for v3), might have been out of date as well. 

But as you said, additional renames that aren't here won't hurt. I think I'll leave them in for those that may have not had such an up-to-date schema as you seem to have in v3.

**Reply by t4sk** [reply]
For GitHub Oauth, I also had to change `providerAccountId` from number to string

```shell

# update github providerAccountId from number to string
db.accounts.updateMany({provider: "github"}, [{$set: {providerAccountId: {$toString: "$providerAccountId"}}}])
```

**Reply by ndom91** [reply]
@t4sk thanks for the Mongo snippet. This was already mentioned in the technical "MySQL" and "Postgres" migration docs (https://next-auth.js.org/getting-started/upgrade-v4#database-migration), however was apparenlty missing from the Mongo one. Would be awesome if you could put up a PR :grin:
